keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md

```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
   - Provision the GMSA and gives details to application admin.
   - Assign access to a AD security group to control what machines can use this GMSA. The AD security group should include all authorized Kubernetes nodes.
 
-2. A Kubernetes admin will create a configmap (in the desired namespace) and populate it with configuration information associated with the GMSA:
-  - Run a Windows PowerShell script to verify the account exists, and capture enough data to uniquely identify the GMSA into a JSON file (credspec). This doesn't contain any passwords or crypto secrets.
+2. A Kubernetes cluster admin (or cluster setup tools or an operator) will install a cluster scoped CRD (of kind GMSACredSpec) whose instances will be used to store GMSA credential spec configuration:
 
-Example credspec.json
 ```
-{
-  "CmsPlugins": [
-    "ActiveDirectory"
-  ],
-  "DomainJoinConfig": {
-    "DnsName": "contoso.com",
-    "Guid": "244818ae-87ca-4fcd-92ec-e79e5252348a",
-    "DnsTreeName": "contoso.com",
-    "NetBiosName": "DEMO",
-    "Sid": "S-1-5-21-2126729477-2524075714-3094792973",
-    "MachineAccountName": "WebApplication1"
-  },
-  "ActiveDirectoryConfig": {
-    "GroupManagedServiceAccounts": [
-      {
-        "Name": "WebApplication1",
-        "Scope": "DEMO"
-      },
-      {
-        "Name": "WebApplication1",
-        "Scope": "contoso.com"
-      }
-    ]
-  }
-}
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: gmsacredspecs.auth.k8s.io
+spec:
+  group: auth.k8s.io
+  version: v1alpha1
+  names:
+    kind: GMSACredSpec
+    plural: gmsacredspecs
+  scope: Cluster
 ```
 
-  - Create a configmap in the namespace with the contents of the JSON - `kubectl create configmap webserver-credspec --from-file=gmsa-cred-spec=/path/to/credspec.json`
+A GMSACredSpec may be used by sidecar containers across different namespaces. Therefore the CRD needs to be cluster scoped.
 
-3. Secure the configmap with the GMSA credspec so that only desired users/service accounts within a namespace can view, modify or use the GMSA configmap:
-  - Create a role to view, modify or use the GMSA configmap:
+3. A Kubernetes cluster admin will create a GMSACredSpec object populated with the credspec configuration associated with a desired GMSA:
+
+  - The cluster admin may run a Windows PowerShell script to generate the YAML definition of a GMSACredSpec object populated with the GMSA credspec details. This doesn't contain any passwords or crypto secrets. Example credspec.yaml for GMSA webapplication1:
+
+```
+apiVersion: auth.k8s.io/v1alpha1
+kind: GMSACredSpec
+metadata:
+  name: "webapp1-credspec"
+credspec:
+  ActiveDirectoryConfig:
+    GroupManagedServiceAccounts:
+    - Name: WebApplication1
+      Scope: DEMO
+    - Name: WebApplication1
+      Scope: contoso.com
+  CmsPlugins:
+  - ActiveDirectory
+  DomainJoinConfig:
+    DnsName: contoso.com
+    DnsTreeName: contoso.com
```

Comment:
why is our definition of the GMSACredSpec different from the details that active directory defines for a GMSA?
the fields in AD for gmsa as returned by `Get-ADServiceAccount -Identity webapplication1` are
1. distringuished name: cn=webapplication1,CN=managed service accounts,CN=contoso,CN=com
2. name = webapplication1
3. SID = <sid>
There is no netbios name or DNS name (those are inferred from the distinguished name)

Reply:
The above is based on the output of `Get-CredentialSpec` as Microsoft recommends in: https://raw.githubusercontent.com/Microsoft/Virtualization-Documentation/live/windows-server-container-tools/ServiceAccounts/CredentialSpec.psm1 
and described in 
https://blogs.msdn.microsoft.com/containerstuff/2017/01/30/create-a-container-with-active-directory-support/
I will add refs to the above to clarify. The only difference from what is generated by the above powershell scripts is I am suggesting the data be stored in YAML to align with everything else in Kubernetes and changed to JSON when necessary as a blob.

Reply:
@ddebroy  thanks for clarifying. it is interesting that they are different in structure, but i am ok with your approach

Reply:
@ddebroy  one last thing. if the scope of the GMSA is more than one node (as defined in active directory), how should that be represented in the yaml? you don't have to put such an example in the KEP, but we should have that in our docs.